make cargo submodule optional

onur-ozkan opened the pull request:
Right now, we fetch the cargo submodule no matter what, even if the command we are running doesn't need it (e.g., `x build compiler library`). This PR changes that to only fetch the cargo submodule when it's necessary.

For more context, see the zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Why.20is.20cargo.20always.20checked.20out.3F

onur-ozkan committed:
bootstrap: exclude cargo from package metadata

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
don't fetch/sync cargo submodule by default

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
refactor `tool_doc` macro in bootstrap

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
handle cargo submodule in a lazy-load way

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
I will wait until #126230 is merged so I don't need to duplicate the tidy part.

RalfJung commented on src/bootstrap/src/core/build_steps/doc.rs:
```diff
@@ -921,6 +920,15 @@ macro_rules! tool_doc {
             fn run(self, builder: &Builder<'_>) {
+                let source_type = SourceType::InTree;
+                $(
+                    let _ = source_type; // silent the "unused variable" warning
```
```suggestion
                    let _ = source_type; // silence the "unused variable" warning
```

ehuss reviewed (commented):
Looks like the doc comment for `workspace_members` should be updated.

I think the checkout is also needed in `tool::Cargo`, otherwise commands like `./x.py test src/tools cargo` or `./x.py build src/tools/cargo` will fail.

It will also need to be added to the vendor command to support `./x.py vendor`.

onur-ozkan commented:
Rebased.

@rustbot ready

Kobzol commented:
r? @Kobzol

Kobzol commented on src/bootstrap/src/core/metadata.rs:
```diff
@@ -89,12 +86,8 @@ fn workspace_members(build: &Build) -> impl Iterator<Item = Package> {
     // Collects `metadata.packages` from all workspaces.
     let packages = collect_metadata("Cargo.toml");
-    let cargo_packages = collect_metadata("src/tools/cargo/Cargo.toml");
     let ra_packages = collect_metadata("src/tools/rust-analyzer/Cargo.toml");
```
While we're at it, shouldn't we also remove Rust analyzer? It also isn't a member of the workspace, right?

Kobzol commented on src/bootstrap/src/core/build_steps/doc.rs:
```diff
@@ -888,12 +888,11 @@ impl Step for Rustc {
-        $(in_tree = $in_tree:literal ,)?
         $(is_library = $is_library:expr,)?
         $(crates = $crates:expr)?
+        $(, submodule $(= $submodule:literal)? )?
```
Why isn't this just an optional argument, rather than a repeating argument?

onur-ozkan replied:
It's already an optional argument, not a repeating one

Kobzol replied:
Sorry :man_facepalming: I misread the macro.

Kobzol reviewed (commented):
Tried all commands that I could think of and everything seemed to work.

onur-ozkan committed:
remove unnecessary packages from `metadata::workspace_members`

Currently bootstrap doesn't use any inner paths from rust-analyzer and
bootstrap with `ShouldRun::create_or_deps`.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

Kobzol commented:
Ok, looks good! It's possible that we have missed something, but the main commands all seem to work, so hopefully it should be fine (if not, we know what to do, sprinkle `cargo` checkout on more places).

Feel free to r=me once CI is green.

onur-ozkan commented:
@bors r=Kobzol

bors commented:
:pushpin: Commit 3457ecc776e87550bc1b1bd4bb686bde8bd5c504 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 3457ecc776e87550bc1b1bd4bb686bde8bd5c504 -->
<!-- homu: {"type":"Approved","sha":"3457ecc776e87550bc1b1bd4bb686bde8bd5c504","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
